// @netlify/build-info/lib/frameworks/docusaurus.js
import { BaseFramework, Category } from './framework.js';
export class Docusaurus extends BaseFramework {
    id = 'docusaurus';
    name = 'Docusaurus';
    configFiles = [
        'docusaurus.config.js',
        'siteConfig.js', // v1
    ];
    npmDependencies = [
        '@docusaurus/core',
        'docusaurus', // v1
    ];
    category = Category.SSG;
    staticAssetsDirectory = 'static';
    dev = {
        command: 'docusaurus start',
        port: 3000,
        pollingStrategies: [{ name: 'TCP' }, { name: 'HTTP' }],
    };
    build = {
        command: 'docusaurus build',
        directory: 'build',
    };
    logo = {
        default: '/logos/docusaurus/default.svg',
        light: '/logos/docusaurus/default.svg',
        dark: '/logos/docusaurus/default.svg',
    };
    env = { BROWSER: 'none' };
    async detect() {
        await super.detect();
        if (this.detected) {
            if (this.isV1(this.detected)) {
                this.build.command = 'docusaurus-build';
                this.build.directory = 'build/<project-name>';
                this.dev.command = 'docusaurus-start';
            }
            return this;
        }
    }
    isV1(detected) {
        return detected.config?.endsWith('siteConfig.js') || detected.package?.name === 'docusaurus';
    }
}
